### PR TITLE
fix(rome_js_syntax): Allow module syntax in cts files

### LINKED ISSUE #3918: 🐛 Illegal use of an export declaration outside of a module for .cts files
### Environment information

```block
vscode extension
```

### What happened?

I installed the vscode extension into a simple project exploring typescript module resolution.

I got an error from rome on a file:

```ts
// example.cts

export function helper() {
    console.log("hello world!");
}
```

As far as I know, this is valid TypeScript code.  In fact, it's taken from the TypeScript [docs](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-rc/#new-file-extensions).  The `.cts` extension indicates that the compilation target should be cjs.  But rome throws an error:

```
Illegal use of an export declaration outside of a module
```

### Expected result

Rome should not error for valid TypeScript code.

### Code of Conduct

- [X] I agree to follow Rome's Code of Conduct

### PR DESCRIPTION
**Pull request by nissy-dev**
## Summary

Fixes #3918

This PR is working in progress. Based on #3958, some regression will happen. I will check them and update the PR if necessary.

## Documentation

- [ ] The PR requires documentation
- [ ] I will create a new PR to update the documentation

### CODE REVIEW

**Comment by nissy-dev** [comment]
I checked these regressions and all of them was happened in the case that `module: node16/nodenext` was specified in tsconfig.json.

If `module: node16/nodenext` was specified in tsconfig.json, ts parser follows Node.js ESM resolution (see: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html). But, Rome doesn't support this resolution logic. Therefore, Rome parser is not reporting errors where it should, resulting in these regressions.

Considering not only linter/formatter but also compiler/bundler, it is better to fix them but I think it is not needed for now. I think there are not so many users to use cts/mts or `module: node16/nodenext`, so It might not be a problem to fix this regression until we get a little more feedback from the community.

File: crates/rome_js_syntax/src/source_type.rs [hunk 1]
```diff
@@ -234,16 +234,16 @@ fn compute_source_type_from_path_or_extension(
...
         SourceType::d_ts()
-    } else if file_name.ends_with(".d.cts") {
-        SourceType::d_ts().with_module_kind(ModuleKind::Script)
     } else {
         match extension {
             "cjs" => SourceType::js_module().with_module_kind(ModuleKind::Script),
             "js" | "mjs" | "jsx" => SourceType::jsx(),
-            "ts" | "mts" => SourceType::ts(),
-            "cts" => SourceType::ts().with_module_kind(ModuleKind::Script),
+            "ts" | "mts" | "cts" => SourceType::ts(),
```
[the remark this answers is not in the document]
Comment (nissy-dev) [reply crates/rome_js_syntax/src/source_type.rs:246]: If keep `with_module_kind(ModuleKind::Script)`, we can't use import/export statements in cts file. TS allows user to use module syntax like import/export statements in cts file. 

see: https://www.typescriptlang.org/docs/handbook/esm-node.html

[the remark this answers is not in the document]
Comment (nissy-dev) [reply crates/rome_js_syntax/src/source_type.rs:246]: And then, this PR is the same as the previous PR (#3958) you approved.

File: xtask/lintdoc/src/main.rs [hunk 2]
```diff
@@ -427,12 +427,9 @@ impl FromStr for CodeBlockTest {
                 "js" | "mjs" | "jsx" => {
                     test.source_type = SourceType::jsx();
                 }
-                "ts" | "mts" => {
+                "ts" | "mts" | "cts" => {
                     test.source_type = SourceType::ts();
                 }
-                "cts" => {
```
[the remark this answers is not in the document]
Comment (nissy-dev) [reply xtask/lintdoc/src/main.rs:442]: This is the same reason as https://github.com/rome/tools/pull/4317#discussion_r1153377586.

File: crates/rome_js_syntax/src/source_type.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_js_syntax/src/source_type.rs:246]: Oooooh I see :O they **transform** during the compilation. Got it!

**Pull request merged**